File: src/node/src/generator/text/textv2.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "textv2.h"

#include <cstdio>
#include <vector>

#include <olive/core/util/cpuoptimize.h>

#include "textbackend.h"

namespace olive
{

#define super ShapeNodeBase

enum TextVerticalAlign {
	k_vertical_align_top,
	k_vertical_align_center,
	k_vertical_align_bottom,
};

const std::string TextGeneratorV2::k_text_input = "text_in";
const std::string TextGeneratorV2::k_html_input = "html_in";
const std::string TextGeneratorV2::k_v_align_input = "valign_in";
const std::string TextGeneratorV2::k_font_input = "font_in";
const std::string TextGeneratorV2::k_font_size_input = "font_size_in";

TextGeneratorV2::TextGeneratorV2()
{
	add_input(k_text_input, NodeValue::k_text, "Sample Text");

	add_input(k_html_input, NodeValue::k_boolean, false);

	add_input(k_v_align_input, NodeValue::k_combo, k_vertical_align_top);

	add_input(k_font_input, NodeValue::k_font);

	add_input(k_font_size_input, NodeValue::k_float, 72.0f);

	set_standard_value(k_color_input, Variant::from_value(Color(1.0f, 1.0f, 1.0)));
	set_standard_value(k_size_input, Vector2D(400, 300));

	set_flag(k_dont_show_in_create_menu);
}

std::string TextGeneratorV2::name() const
{
	return "Text (Legacy)";
}

std::string TextGeneratorV2::id() const
{
	return "org.olivevideoeditor.Olive.text2";
}

std::vector<Node::CategoryID> TextGeneratorV2::category() const
{
	return { k_category_generator };
}

std::string TextGeneratorV2::description() const
{
	return "Generate rich text.";
}

void TextGeneratorV2::retranslate()
{
	super::retranslate();

	set_input_name(k_text_input, "Text");
	set_input_name(k_html_input, "Enable HTML");
	set_input_name(k_font_input, "Font");
	set_input_name(k_font_size_input, "Font Size");
	set_input_name(k_v_align_input, "Vertical Align");
	set_combo_box_strings(k_v_align_input, { "Top", "Center", "Bottom" });
}

void TextGeneratorV2::value(const NodeValueRow &value,
							const NodeGlobals &globals,
							NodeValueTable *table) const
{
	if (!value.at(k_text_input).to_string().empty()) {
		GenerateJob job(value);
		auto text_params = globals.vparams();
		text_params.set_format(PixelFormat::f32);
		table->push(NodeValue::k_texture, Texture::job(text_params, job), this);
	}
}

void TextGeneratorV2::generate_frame(FramePtr frame,
									const GenerateJob &job) const
{
	// Formerly laid out and rasterized with QTextDocument + QPainter into a
	// grayscale QImage (the alpha channel was then transplanted to the float
	// buffer with the correct RGB). Layout/rasterization now runs behind the
	// facade-installed text backend (see textbackend.h); the coverage buffer
	// handling and the transplant loop are unchanged.
	const int width = frame->width();
	const int height = frame->height();

	// QImage::Format_Grayscale8 scanlines are 32-bit aligned
	const int linesize = (width + 3) & ~3;
	std::vector<unsigned char> img(size_t(linesize) * height, 0);

	TextLayoutRequest req;
	req.text = job.get(k_text_input).to_string();
	req.font_family = job.get(k_font_input).to_string();
	req.font_size_pt =
		job.get(k_font_size_input).to_double();
	// 72 DPI in DPM (72 / 2.54 * 100)
	req.dots_per_meter = 2835;
	if (job.get(k_html_input).to_bool()) {
		req.mode = TextLayoutRequest::k_html;
		// QTextDocument::setHtml() doesn't translate newlines, so they were
		// replaced with <br> tags first
		std::string::size_type nl = 0;
		while ((nl = req.text.find('\n', nl)) != std::string::npos) {
			req.text.replace(nl, 1, "<br>");
			nl += 4;
		}
	}

	Vector2D size = job.get(k_size_input).to_vec2();
	req.wrap_width = size.x();

	TextLayoutSize doc;
	if (TextMeasureBackend measure = text_measure_backend()) {
		doc = measure(req);
	}

	double scale = 1.0 / frame->video_params().divider();

	Vector2D pos = job.get(k_position_input).to_vec2();
	double base_offset_x = pos.x() - size.x() / 2 +
						   frame->video_params().width() / 2;
	double base_offset_y = pos.y() - size.y() / 2 +
						   frame->video_params().height() / 2;

	TextVerticalAlign valign = static_cast<TextVerticalAlign>(
		job.get(k_v_align_input).to_int());
	int doc_height = int(doc.height);

	double draw_offset_x = base_offset_x;
	double draw_offset_y = base_offset_y;

	switch (valign) {
	case k_vertical_align_top:
		// Do nothing
		break;
	case k_vertical_align_center:
		// Center align
		draw_offset_y += size.y() / 2 - doc_height / 2;
		break;
	case k_vertical_align_bottom:
		draw_offset_y += size.y() - doc_height;
		break;
	}

	if (TextRenderBackend render = text_render_backend()) {
		TextRenderTransform t;
		t.scale = scale;
		t.draw_offset_x = draw_offset_x;
		t.draw_offset_y = draw_offset_y;
		// The clip rect was set before the vertical alignment translate
		t.clip_enabled = true;
		t.clip_offset_x = base_offset_x;
		t.clip_offset_y = base_offset_y;
		t.clip_width = size.x();
		t.clip_height = size.y();
		TextRenderTarget target = { img.data(), width, height, linesize, 1 };
		render(req, t, target);
	} else {
		static bool warned = false;
		if (!warned) {
			fprintf(stderr,
					"TextGeneratorV2: no text backend installed, frame left empty\n");
			warned = true;
		}
	}

	// Transplant alpha channel to frame
	Color rgba = job.get(k_color_input).to_color();
#if defined(OLIVE_PROCESSOR_X86) || defined(OLIVE_PROCESSOR_ARM)
	__m128 sse_color = _mm_loadu_ps(rgba.data());
#endif

	float *frame_dst = reinterpret_cast<float *>(frame->data());
	for (int y = 0; y < height; y++) {
		const unsigned char *src_y = img.data() + size_t(linesize) * y;
		float *dst_y = frame_dst + y * frame->linesize_pixels() *
									   VideoParams::k_rgba_channel_count;

		for (int x = 0; x < width; x++) {
			float alpha = float(src_y[x]) / 255.0f;
			float *dst = dst_y + x * VideoParams::k_rgba_channel_count;

#if defined(OLIVE_PROCESSOR_X86) || defined(OLIVE_PROCESSOR_ARM)
			__m128 sse_alpha = _mm_load1_ps(&alpha);
			__m128 sse_res = _mm_mul_ps(sse_color, sse_alpha);

			_mm_store_ps(dst, sse_res);
#else
			for (int i = 0; i < VideoParams::k_rgba_channel_count; i++) {
				dst[i] = rgba.data()[i] * alpha;
			}
#endif
		}
	}
}

}

File: src/node/rust/src/nodes/textv2.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Legacy text generator v2 (C++
//! `src/node/src/generator/text/textv2.{h,cpp}`,
//! `olive::TextGeneratorV2`, derives from `ShapeNodeBase`).
//!
//! FONT/RASTER BACKEND DEPENDENCY — DELIBERATELY UNDECIDED:
//! The C++ does NOT link any font/raster library directly (no freetype,
//! stb, harfbuzz, etc. anywhere in the tree). Text layout/rasterization
//! was historically Qt's rich text stack (`QTextDocument` +
//! `QAbstractTextDocumentLayout` + `QPainter` into a
//! `QImage::Format_Grayscale8` coverage buffer); it now runs behind the
//! facade-installed hooks in [`super::textbackend`]. No Rust font crate
//! is chosen here on purpose.

use crate::factory::NodeMeta;
use crate::node::{Category, NodeBehavior, NodeCore};

/// Text input id (C++ `k_text_input`). Type: text; default
/// `"Sample Text"`.
pub const TEXT_INPUT: &str = "text_in";

/// HTML toggle input id (C++ `k_html_input`). Type: boolean; default
/// `false`.
pub const HTML_INPUT: &str = "html_in";

/// Vertical alignment input id (C++ `k_v_align_input`). Type: combo;
/// default `0` (C++ `k_vertical_align_top`); combo strings: "Top",
/// "Center", "Bottom".
pub const V_ALIGN_INPUT: &str = "valign_in";

/// Font family input id (C++ `k_font_input`). Type: font (C++
/// `NodeValue::k_font` — no dedicated [`crate::value::ValueType`]
/// variant, travels as text); no default.
pub const FONT_INPUT: &str = "font_in";

/// Font size input id (C++ `k_font_size_input`). Type: float; default
/// `72.0`.
pub const FONT_SIZE_INPUT: &str = "font_size_in";

/// Legacy text generator v2. The C++ class has no own private members
/// — it inherits position/size/color inputs and the polygon gizmo from
/// the shape base (C++ `ShapeNodeBase`, to be modelled in
/// `super::shapenodebase`) and everything else lives in [`NodeCore`] —
/// so this is a unit struct.
pub struct TextGeneratorV2;

impl NodeBehavior for TextGeneratorV2 {
	/// Human-readable name (C++ `name()`).
	fn name(&self) -> &str {
		"Text (Legacy)"
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.text2"
	}

	/// Categories (C++ `category()`).
	fn categories(&self) -> &[Category] {
		&[Category::Generator]
	}

	/// Description (C++ `description()`).
	fn description(&self) -> &str {
		"Generate rich text."
	}

	/// Localized input names (C++ `retranslate()`): `text_in` ->
	/// "Text", `html_in` -> "Enable HTML", `font_in` -> "Font",
	/// `font_size_in` -> "Font Size", `valign_in` -> "Vertical Align"
	/// (combo strings Top/Center/Bottom); the base class retranslate
	/// covers the inherited shape inputs.
	fn input_name<'a>(&self, id: &'a str) -> &'a str {
		todo!()
	}

	/// Evaluate outputs (C++ `value()`): if the text input is
	/// non-empty, push a texture generate job at the global video
	/// params forced to `PixelFormat::f32`; otherwise push nothing.
	fn value(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		time: oakcore_rs::Rational,
		table: &mut crate::value::NodeValueTable,
	) {
		todo!()
	}

	/// Direct frame generation (C++ `generate_frame()`): rasterizes
	/// the text into a grayscale coverage buffer via the installed
	/// text backend at 72 DPI (2835 dots/meter) — HTML mode replaces
	/// newlines with `<br>`, wrap width is the shape size X, base
	/// offset is the shape position re-centered into frame space,
	/// vertical draw offset depends on the valign combo (top: none;
	/// center: `size.y/2 - doc.height/2`; bottom: `size.y -
	/// doc.height`), and the clip rect is the shape rect at the base
	/// offset — then writes the coverage multiplied by the RGBA color
	/// into the float frame (SIMD path in C++, scalar fallback
	/// identical). With no backend installed, warns once and leaves
	/// the frame empty.
	fn generate_frame(
		&self,
		core: &NodeCore,
		frame: &mut crate::bridge::render::TextureHandle,
		time: oakcore_rs::Rational,
	) {
		todo!()
	}

	/// Deep copy (C++ `copy()`).
	fn duplicate(&self, core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		todo!()
	}
}

/// Constructor (C++ `TextGeneratorV2::TextGeneratorV2()`): adds
/// `text_in`, `html_in`, `valign_in`, `font_in` and `font_size_in` with
/// the defaults documented on the constants, sets the inherited shape
/// base `color_in` standard value to white and `size_in` to
/// `(400, 300)`, and sets the `dont_show_in_create_menu` flag.
pub fn create() -> (NodeCore, Box<dyn NodeBehavior>) {
	todo!()
}

/// Register this node type (C++ `k_text_generator_v2` in
/// `factory.cpp::create_from_factory_index`).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.text2",
		name: "Text (Legacy)",
		categories: &[Category::Generator],
		create,
	});
}
